src/main/kotlin/gg/essential/universal/render/URenderPipeline.kt

```diff
@@ -0,0 +1,529 @@
+            //$$             withFragmentShader(shader.frag)
+            //$$
+            //$$             shader.samplers.forEach { withSampler(it) }
+            //$$             shader.uniforms.forEach { withUniform(it.key, it.value) }
+            //$$         }
+            //$$     }
+            //$$     withDepthTestFunction(when (depthTest) {
+            //$$         DepthTest.Disabled -> DepthTestFunction.NO_DEPTH_TEST
+            //$$         DepthTest.Always -> DepthTestFunction.NO_DEPTH_TEST // implemented via raw GlStateManager below
+            //$$         DepthTest.Equal -> DepthTestFunction.EQUAL_DEPTH_TEST
+            //$$         DepthTest.LessOrEqual -> DepthTestFunction.LEQUAL_DEPTH_TEST
+            //$$         DepthTest.Less -> DepthTestFunction.LESS_DEPTH_TEST
+            //$$         DepthTest.Greater -> DepthTestFunction.GREATER_DEPTH_TEST
+            //$$     })
+            //$$     withCull(culling)
+            //$$     withColorLogic(when (colorLogic) {
+            //$$         ColorLogic.None -> LogicOp.NONE
+            //$$         ColorLogic.OrReverse -> LogicOp.OR_REVERSE
+            //$$     })
+            //$$     val blendState = blendState
+            //$$     if (blendState.enabled) {
+            //$$         withBlend(BlendFunction(
+            //$$             blendState.srcRgb.mcSourceFactor,
+            //$$             blendState.dstRgb.mcDestFactor,
+            //$$             blendState.srcAlpha.mcSourceFactor,
+            //$$             blendState.dstAlpha.mcDestFactor,
+            //$$         ))
+            //$$     } else {
+            //$$         withoutBlend()
+            //$$     }
+            //$$     colorMask.let { (colorMask, alphaMask) ->
+            //$$         withColorWrite(colorMask, alphaMask)
+            //$$     }
+            //$$     withDepthWrite(depthMask)
+            //$$     polygonOffset.let { (factor, units) ->
+            //$$         withDepthBias(factor, units)
+            //$$     }
+            //$$ }.build()
+            //$$
+            //$$ if (depthTest == DepthTest.Always) {
+            //$$     abstract class CustomRenderPipeline(inner: RenderPipeline) : RenderPipeline(inner.location, inner.vertexShader, inner.fragmentShader, inner.shaderDefines, inner.samplers, inner.uniforms, inner.blendFunction, inner.depthTestFunction, inner.polygonMode, inner.isCull, inner.isWriteColor, inner.isWriteAlpha, inner.isWriteDepth, inner.colorLogic, inner.vertexFormat, inner.vertexFormatMode, inner.depthBiasScaleFactor, inner.depthBiasConstant)
+            //$$     mcRenderPipeline = object : CustomRenderPipeline(mcRenderPipeline) {
+            //$$         val stackWalker = StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE)
+            //$$         override fun isCull(): Boolean {
+            //$$             if (stackWalker.callerClass == GlResourceManager::class.java) {
```

Comment:
I can't say I love this, but I assume that without mixin we can't do any better.